### PR TITLE
Stop deleting DaemonSet pods since they will be recreated immediately, and also require a flag when there are DaemonSet pods on the node.

### PR DESCRIPTION
**Pull request by mml**
It took some work to get all the cases right in terms of error and warning messages, but the transcript below shows that it's good.  I will refactor that part and squash the commits, but putting the PR out for a little scrutiny of semantics.

``` console
% echo $m
kubernetes-minion-c9db
% _output/local/bin/linux/amd64/kubectl drain $m
node "kubernetes-minion-c9db" cordoned
error: pods managed by neither a ReplicationController nor a DaemonSet: naked (use --force to override) and DaemonSet-managed pods: mmlds-0w8zz (use --ignore-daemonsets to ignore)

% _output/local/bin/linux/amd64/kubectl drain $m --force
node "kubernetes-minion-c9db" already cordoned
error: DaemonSet-managed pods: mmlds-0w8zz (use --ignore-daemonsets to ignore)

% _output/local/bin/linux/amd64/kubectl drain $m --ignore-daemonsets
node "kubernetes-minion-c9db" already cordoned
error: pods managed by neither a ReplicationController nor a DaemonSet: naked (use --force to override)

% _output/local/bin/linux/amd64/kubectl drain $m --force --ignore-daemonsets
node "kubernetes-minion-c9db" already cordoned
WARNING: About to delete these pods managed by neither a ReplicationController nor a DaemonSet: naked
pod "naked" deleted
pod "kube-dns-v10-luuw6" deleted
node "kubernetes-minion-c9db" drained
```

### CODE REVIEW

**Comment by mml** [comment]
@davidopp

**Comment by davidopp** [comment]
Yeah, figuring out the right semantics here is tough, I think what you chose is pretty reasonable (I can think of other semantics, but none are obviously better). 

I have two suggestions
(1) maybe call it "--nowarn-daemonsets" instead of "--ignore-daemonsets" since I think "nowarn" is a little clearer about what the flag does. 
(2) when all the requirements pass to do a drain and you actually do it (e.g. last case in your sample output above), print "skipping DaemonSet XXX" whenever there is a DaemonSet on the node, to make it a little clearer what is going on.

Does that sound reasonable?

**Comment by mml** [comment]
> I have two suggestions
> (1) maybe call it "--nowarn-daemonsets" instead of "--ignore-daemonsets" since I think "nowarn" is a little clearer about what the flag does. 

The flag doesn't just turn off a warning.  In fact, without it we won't drain a node with a DaemonSet pod.  With it, we literally skip over those pods, not deleting them but still saying "drained".  There may be a better name, but "nowarn" makes it sound like it only toggles diagnostics and not behavior.

> (2) when all the requirements pass to do a drain and you actually do it (e.g. last case in your sample output above), print "skipping DaemonSet " whenever there is a DaemonSet on the node (one for each DaemonSet you find), to make it a little clearer what is going on.

This I can add.

**Comment by davidopp** [comment]
> The flag doesn't just turn off a warning. In fact, without it we won't drain a node with a DaemonSet pod. With it, we literally skip over those pods, not deleting them but still saying "drained". There may be a better name, but "nowarn" makes it sound like it only toggles diagnostics and not behavior.

Ah yes, I was just thinking about the fact that we never actually kill DaemonSet pods under any circumstances, hence the flag's behavior with respect to DaemonSet pods is really just a warning. But you're right, it changes the behavior with respect to machines (i.e. whether we drain other pods on the machine or skip the machine). So current flag name sounds good.

**Comment by mml** [comment]
@davidopp here's the new output.

```console
% _output/local/bin/linux/amd64/kubectl drain kubernetes-minion-c9db --ignore-daemonsets --force
node "kubernetes-minion-c9db" already cordoned
WARNING: About to delete these pods managed by neither a ReplicationController nor a DaemonSet: naked and DaemonSet-managed pods: mmlds-l1ekr
pod "naked" deleted
pod "kube-dns-v10-sp51v" deleted
node "kubernetes-minion-c9db" drained
```

**Comment by mml** [comment]
NM.  Just realized I printed exactly the wrong thing, fixing it now.

**Comment by mml** [comment]
```
WARNING: About to delete these pods managed by neither a ReplicationController nor a DaemonSet: naked
WARNING: Skipping DaemonSet-managed pods: mmlds-bmee1
pod "naked" deleted
pod "kube-dns-v10-rlohl" deleted
node "kubernetes-minion-c9db" drained
```

**Comment by mml** [comment]
@davidopp PTAL

File: docs/man/man1/kubectl-drain.1 [hunk 1]
```diff
@@ -18,9 +18,11 @@ Drain node in preparation for maintenance.
 .PP
 The given node will be marked unschedulable to prevent new pods from arriving.
 Then drain deletes all pods except mirror pods (which cannot be deleted through
-the API server).  If there are any pods that are neither mirror pods nor
-managed by a ReplicationController or DaemonSet, then drain will not delete any
-pods unless you use \-\-force.
+the API server) and DaemonSet\-managed pods.  If there are DaemonSet\-managed
+pods, drain will not proceed without \-\-ignore\-daemonsets, since those pods will
+be immediately replaced by the NodeController.  If there are any pods that are
```
Comment (davidopp) [inline docs/man/man1/kubectl-drain.1]: we never actually moved this to the NodeController; the daemons would be replaced by the DaemonSet controller.

File: docs/user-guide/kubectl/kubectl_drain.md [hunk 2]
```diff
@@ -38,9 +38,11 @@ Drain node in preparation for maintenance.
 
 The given node will be marked unschedulable to prevent new pods from arriving.
 Then drain deletes all pods except mirror pods (which cannot be deleted through
-the API server).  If there are any pods that are neither mirror pods nor
-managed by a ReplicationController or DaemonSet, then drain will not delete any
-pods unless you use --force.
+the API server) and DaemonSet-managed pods.  If there are DaemonSet-managed
+pods, drain will not proceed without --ignore-daemonsets, since those pods will
+be immediately replaced by the NodeController.  If there are any pods that are
```
Comment (davidopp) [inline docs/user-guide/kubectl/kubectl_drain.md]: DaemonSet controller

File: pkg/kubectl/cmd/drain.go [hunk 3]
```diff
@@ -98,9 +100,11 @@ const (
 
 The given node will be marked unschedulable to prevent new pods from arriving.
 Then drain deletes all pods except mirror pods (which cannot be deleted through
-the API server).  If there are any pods that are neither mirror pods nor
-managed by a ReplicationController or DaemonSet, then drain will not delete any
-pods unless you use --force.
+the API server) and DaemonSet-managed pods.  If there are DaemonSet-managed
+pods, drain will not proceed without --ignore-daemonsets, since those pods will
+be immediately replaced by the NodeController.  If there are any pods that are
```
Comment (davidopp) [inline pkg/kubectl/cmd/drain.go]: DaemonSet controller

File: pkg/kubectl/cmd/drain.go [hunk 4]
```diff
@@ -227,28 +234,65 @@ func (o *DrainOptions) getPodsForDeletion() ([]api.Pod, error) {
 				// gone/missing, not for any other cause.  TODO(mml): something more
 				// sophisticated than this
 				if err == nil && ds != nil {
-					replicated = true
+					// Otherwise, treat daemonset-managed pods as unmanaged since
+					// NodeController currently ignores the unschedulable bit.
```
Comment (davidopp) [inline pkg/kubectl/cmd/drain.go]: DaemonSet controller

File: pkg/kubectl/cmd/drain.go [hunk 5]
```diff
@@ -227,28 +234,65 @@ func (o *DrainOptions) getPodsForDeletion() ([]api.Pod, error) {
...
+		if o.Force {
+			return pods, errors.New(unmanagedMsg([]string{}, daemonSetPodNames, true))
+		} else {
+			return pods, errors.New(unmanagedMsg(unreplicatedPodNames, daemonSetPodNames, true))
+		}
+	} else if len(unreplicatedPodNames) > 0 {
+		// If we get here, we know o.Force is true.
+		fmt.Fprintf(o.out, "WARNING: About to delete these %s\n", unmanagedMsg(unreplicatedPodNames, []string{}, false))
+		if len(daemonSetPodNames) > 0 {
+			fmt.Fprintf(o.out, "WARNING: Skipping %s\n", unmanagedMsg([]string{}, daemonSetPodNames, false))
```
Comment (davidopp) [inline pkg/kubectl/cmd/drain.go]: I think we also want to print this message in the case where unreplicatedPodNames == 0 && o.IgnoreDaemonsets && len(daemonSetPodNames) > 0, but if I'm understanding the logic correctly, you won't do that.

**Comment by mml** [comment]
> I think we also want to print this message in the case where unreplicatedPodNames == 0 && o.IgnoreDaemonsets && len(daemonSetPodNames) > 0, but if I'm understanding the logic correctly, you won't do that.

Correct.  Fixed.  Also removed all mention of NodeController in favor of DaemonSet Controller.  PTAL

File: docs/man/man1/kubectl-drain.1 [hunk 6]
```diff
@@ -18,9 +18,11 @@ Drain node in preparation for maintenance.
 .PP
 The given node will be marked unschedulable to prevent new pods from arriving.
 Then drain deletes all pods except mirror pods (which cannot be deleted through
-the API server).  If there are any pods that are neither mirror pods nor
-managed by a ReplicationController or DaemonSet, then drain will not delete any
-pods unless you use \-\-force.
+the API server) and DaemonSet\-managed pods.  If there are DaemonSet\-managed
+pods, drain will not proceed without \-\-ignore\-daemonsets, since those pods will
```
Comment (davidopp) [inline docs/man/man1/kubectl-drain.1]: The "those pods will be immediately replaced by the DaemonSet Controller" thing is a little confusing because drain actually won't delete the DaemonSet pods, even if you pass --ignore-daemonsets. I would rephrase as something like "If there are DaemonSet-managed pods, drain will not proceed without --ignore-daemonsets, and regardless it will not delete any DaemonSet-managed pods, because those pods would be immediately replaced by the DaemonSet controller, which ignores unschedulable markings."

Of course for all of this stuff we need to make sure your Job PR merges properly, i.e. ReplicationController -> ReplicationController, Job

File: pkg/kubectl/cmd/drain.go [hunk 7]
```diff
@@ -98,9 +100,11 @@ const (
 
 The given node will be marked unschedulable to prevent new pods from arriving.
 Then drain deletes all pods except mirror pods (which cannot be deleted through
-the API server).  If there are any pods that are neither mirror pods nor
-managed by a ReplicationController or DaemonSet, then drain will not delete any
-pods unless you use --force.
+the API server) and DaemonSet-managed pods.  If there are DaemonSet-managed
```
Comment (davidopp) [inline pkg/kubectl/cmd/drain.go]: same comment here

File: pkg/kubectl/cmd/drain.go [hunk 8]
```diff
@@ -227,28 +234,65 @@ func (o *DrainOptions) getPodsForDeletion() ([]api.Pod, error) {
...
-		if !o.Force {
-			return pods, fmt.Errorf("refusing to continue due to pods managed by neither a ReplicationController nor a DaemonSet: %s (use --force to override)", joined)
+	if (!o.IgnoreDaemonsets && len(daemonSetPodNames) > 0) || (!o.Force && len(unreplicatedPodNames) > 0) {
+		if o.Force {
+			return pods, errors.New(unmanagedMsg([]string{}, daemonSetPodNames, true))
+		} else {
+			return pods, errors.New(unmanagedMsg(unreplicatedPodNames, daemonSetPodNames, true))
 		}
-		fmt.Fprintf(o.out, "WARNING: About to delete these pods managed by neither a ReplicationController nor a DaemonSet: %s\n", joined)
+	} else if len(unreplicatedPodNames) > 0 {
```
Comment (davidopp) [inline pkg/kubectl/cmd/drain.go]: else not necessary

Comment (davidopp) [inline pkg/kubectl/cmd/drain.go]: sorry, ignore that

File: pkg/kubectl/cmd/drain.go [hunk 9]
```diff
@@ -227,28 +234,65 @@ func (o *DrainOptions) getPodsForDeletion() ([]api.Pod, error) {
...
+		default:
+			pods = append(pods, pod)
 		}
 	}
 
-	if len(unreplicatedPodNames) > 0 {
-		joined := strings.Join(unreplicatedPodNames, ", ")
-		if !o.Force {
-			return pods, fmt.Errorf("refusing to continue due to pods managed by neither a ReplicationController nor a DaemonSet: %s (use --force to override)", joined)
+	if (!o.IgnoreDaemonsets && len(daemonSetPodNames) > 0) || (!o.Force && len(unreplicatedPodNames) > 0) {
```
Comment (davidopp) [inline pkg/kubectl/cmd/drain.go]: Honestly, this logic has just become too complicated for a reader to easily understand. Can you just write a single 16-arm if/else if clause that covers all the possible values of len(daemonSetPodNames), len(unreplicatedPodNames), o.IgnoreDaemonsets, and o.Force?

Also, in the cases where you return error as second argument, can you return empty pod list to make it clearer that it means no pods will be deleted?

**Comment by bgrant0607** [comment]
Have we effectively reverted #17318 -- will daemons land on unschedulable nodes?

**Comment by davidopp** [comment]
> Have we effectively reverted #17318 -- will daemons land on unschedulable nodes?

Yes, it's one of the changes in #20061 which is LGTMd and waiting to merge.

File: pkg/kubectl/cmd/drain.go (hunk 9 above)
Comment (mml) [inline pkg/kubectl/cmd/drain.go]: Good feedback.  I hated this part anyway.  Rewriting it now.

**Comment by mml** [comment]
I didn't quite do 16 cases, but I think it's pretty clear now.  LMK

File: pkg/kubectl/cmd/drain.go [hunk 10]
```diff
@@ -240,24 +252,63 @@ func (o *DrainOptions) getPodsForDeletion() ([]api.Pod, error) {
...
+	}
+
 	return pods, nil
 }
 
+// Helper for generating errors or warnings about unmanaged pods.
+func unmanagedMsg(unreplicatedNames []string, daemonSetNames []string, include_guidance bool) string {
+	msgs := []string{}
+	if len(unreplicatedNames) > 0 {
+		msg := fmt.Sprintf("pods not managed by ReplicationController, Job, or DaemonSet: %s", strings.Join(unreplicatedNames, ","))
```
Comment (davidopp) [inline pkg/kubectl/cmd/drain.go]: This should not say DaemonSet, right?

Comment (mml) [inline pkg/kubectl/cmd/drain.go]: Wrong.  These are pods with no created-by ref, so they aren't managed by RC, Job, or DS.

Comment (davidopp) [inline pkg/kubectl/cmd/drain.go]: Oh yeah, the thing I missed was that you only append to unreplicatedPodNames in a case branch that comes after daemonset_pods, thus daemonset is never in unreplicatedPodNames.

**Comment by davidopp** [comment]
LGTM
